qiskit/dagcircuit/collect_blocks.py

```diff
@@ -254,22 +255,37 @@ def collapse_to_operation(self, blocks, collapse_fn):
         then uses collapse_fn to collapse this circuit into a single operation.
         """
         global_index_map = {wire: idx for idx, wire in enumerate(self.dag.qubits)}
+        global_index_map.update({wire: idx for idx, wire in enumerate(self.dag.clbits)})
+
         for block in blocks:
-            # Find the set of qubits used in this block (which might be much smaller than
-            # the set of all qubits).
+            # Find the sets of qubits/clbits used in this block (which might be much smaller
+            # than the set of all qubits/clbits).
             cur_qubits = set()
+            cur_clbits = set()
+
             for node in block:
                 cur_qubits.update(node.qargs)
+                cur_clbits.update(_get_node_cargs(node))
 
-            # For reproducibility, order these qubits compatibly with the global order.
+            # For reproducibility, order these qubits/clbits compatibly with the global order.
             sorted_qubits = sorted(cur_qubits, key=lambda x: global_index_map[x])
+            sorted_clbits = sorted(cur_clbits, key=lambda x: global_index_map[x])
 
             # Construct a quantum circuit from the nodes in the block, remapping the qubits.
             wire_pos_map = {qb: ix for ix, qb in enumerate(sorted_qubits)}
-            qc = QuantumCircuit(len(cur_qubits))
+            wire_pos_map.update({qb: ix for ix, qb in enumerate(sorted_clbits)})
+
+            qc = QuantumCircuit(sorted_qubits, sorted_clbits)
             for node in block:
-                remapped_qubits = [wire_pos_map[qarg] for qarg in node.qargs]
-                qc.append(CircuitInstruction(node.op, remapped_qubits, node.cargs))
+                instructions = qc.append(CircuitInstruction(node.op, node.qargs, node.cargs))
+                cond = getattr(node.op, "condition", None)
+                if cond is not None:
+                    if not isinstance(cond[0], Clbit):
+                        raise DAGCircuitError(
+                            "Conditional gates must be conditioned on individual Clbits, "
+                            "not ClassicalRegisters to be able to collapse."
+                        )
```

Comment:
Thanks! This looks good. I have moved the code a bit to avoid multiple iterations over registers, done in [db71bbe](https://github.com/Qiskit/qiskit-terra/pull/9823/commits/db71bbe961bcb967865a01af313d44b6bfa23542).